src/proto/metrics2.proto

```diff
@@ -11,72 +11,137 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-syntax = "proto2";
+// This is copied and lightly edited from
+// github.com/prometheus/client_model/io/prometheus/client/metrics.proto
+// and finally converted to proto3 syntax to make it usable for the
+// gogo-protobuf approach taken within prometheus/prometheus.
+
+syntax = "proto3";
 
 package io.prometheus.client;
-option java_package = "io.prometheus.client";
+option go_package = "io_prometheus_client";
+
+//import "google/protobuf/gogo.proto";
+import "google/protobuf/timestamp.proto";
 
 message LabelPair {
-  optional string name  = 1;
-  optional string value = 2;
+  string name  = 1;
+  string value = 2;
 }
 
 enum MetricType {
-  COUNTER    = 0;
-  GAUGE      = 1;
-  SUMMARY    = 2;
-  UNTYPED    = 3;
-  HISTOGRAM  = 4;
+  // COUNTER must use the Metric field "counter".
+  COUNTER = 0;
+  // GAUGE must use the Metric field "gauge".
+  GAUGE = 1;
+  // SUMMARY must use the Metric field "summary".
+  SUMMARY = 2;
+  // UNTYPED must use the Metric field "untyped".
+  UNTYPED = 3;
+  // HISTOGRAM must use the Metric field "histogram".
+  HISTOGRAM = 4;
+  // GAUGE_HISTOGRAM must use the Metric field "histogram".
+  GAUGE_HISTOGRAM = 5;
 }
 
 message Gauge {
-  optional double value = 1;
+  double value = 1;
 }
 
 message Counter {
-  optional double value = 1;
+  double   value    = 1;
+  Exemplar exemplar = 2;
 }
 
 message Quantile {
-  optional double quantile = 1;
-  optional double value    = 2;
+  double quantile = 1;
+  double value    = 2;
 }
 
 message Summary {
-  optional uint64   sample_count = 1;
-  optional double   sample_sum   = 2;
-  repeated Quantile quantile     = 3;
+  uint64            sample_count = 1;
+  double            sample_sum   = 2;
+  repeated Quantile quantile     = 3 ;// [(gogoproto.nullable) = false];
 }
 
 message Untyped {
-  optional double value = 1;
+  double value = 1;
 }
 
 message Histogram {
-  optional uint64 sample_count = 1;
-  optional double sample_sum   = 2;
-  repeated Bucket bucket       = 3; // Ordered in increasing order of upper_bound, +Inf bucket is optional.
+  uint64 sample_count       = 1;
+  double sample_count_float = 4; // Overrides sample_count if > 0.
+  double sample_sum         = 2;
```

Comment:
Will messages sent using this histogram type be compatible with old servers that used the old histogram type? Not that I care, it's up to Prometheus to maintain compatibility, just curious.

Reply:
They will not be because of the change in the metrics definitions. This means that if someone enables the protobof on the seastar side, they should use Prometheus 2.0 version or higher.